src/diagonal.jl

```diff
@@ -1024,8 +1024,8 @@ function svd(D::Diagonal{T}) where {T<:Number}
     Vt = copy(U)
     for i in 1:length(d)
         j = piv[i]
-        U[j,i] = d[j] / S[i]
-        Vt[i,j] = one(Td)
+        U[j,i] = iszero(d[j]) ? oneunit(Td) : d[j] / S[i]
```

Comment:
These should probably be `one(Td)` and not `oneunit(Td)`? The `U` and `V` matrices scale the diagonal entries, and the proper identity element for that operation is given by `one`.

Reply:
That's what I would have thought. But the `eltype` of `V` and `U` is really `oneunit(Td)`, not `one(Td)`. Furthermore, this is explicitly tested for [here](https://github.com/JuliaLang/LinearAlgebra.jl/blob/bb5f566534c669c88a804fb38e6a9997fffa4d2d/test/diagonal.jl#L481).

Reply:
Huh, strange. In that case, this change is fine in this PR, and we may discuss the issue separately.

Reply:
> That's what I would have thought. But the `eltype` of `V` and `U` is really `oneunit(Td)`, not `one(Td)`. Furthermore, this is explicitly tested for [here](https://github.com/JuliaLang/LinearAlgebra.jl/blob/bb5f566534c669c88a804fb38e6a9997fffa4d2d/test/diagonal.jl#L481).

I think I touched this method last. The eltype is `Td`, which is ideally "unitless" (`oneunit(T)/oneunit(T)`). I just had no better idea of how to strip the unit. Currently, this works in the tests by promoting unitless numbers to the unitless version of the "unitful" type `Furlong{0}`.

The better and expected option would be to define `Td = typeof(one(T))`. Then this would get rid of the unit on `U` and `V`, and the two mentioned tests (for `U` and `V`) would need to be modified, and we could use `one(Td)` again.

Reply:
Perhaps we could do the same for the `eigen` method changed in https://github.com/JuliaLang/LinearAlgebra.jl/commit/b62b018ab25b53e85c79b7e5584bbf03bc866143: add a method for `Number` eltypes, and make sure the vectors are truly unitless.